src/main/java/com/iota/iri/service/tipselection/impl/ReferenceCheckerImpl.java

```diff
@@ -0,0 +1,54 @@
+package com.iota.iri.service.tipselection.impl;
+
+import com.iota.iri.controllers.TransactionViewModel;
+import com.iota.iri.model.Hash;
+import com.iota.iri.service.tipselection.ReferenceChecker;
+import com.iota.iri.storage.Tangle;
+import org.apache.commons.collections4.CollectionUtils;
+
+import java.util.*;
+
+public class ReferenceCheckerImpl implements ReferenceChecker {
+    private Tangle tangle;
+
+    public ReferenceCheckerImpl(Tangle tangle) {
+        this.tangle = tangle;
+    }
+
+    @Override
+    public boolean doesReference(Hash referencer, Hash target) throws Exception {
+        if (!TransactionViewModel.exists(tangle, referencer) ||
+            !TransactionViewModel.exists(tangle, target)) {
+                throw new Exception("Transactions not found");
+        }
+
+        LinkedHashSet<Hash> visitedHashes = new LinkedHashSet<>();
+        Queue<Hash> queue = new LinkedList<Hash>();
+
+        // Run BFS scan to look for approvee

```

Comment:
did you look at the DAG Util? if so, what was missing there in order for you to reimplement the algorithm?
(it's a new class so it's an honest question)

Reply:
Was not aware of it, I'll have a look